### ISSUE TITLE
More info about how this compares to auto-complete

### ISSUE THREAD

**Issue by expez** [company-mode/company-mode#68]
Hi, 
it would be nice if if you provided a bit more information about why one might switch from AC to Company, or even pick Company in the first place.

Are they at feature parity?  E.g. in the screenshots company is only shown listing completion candidates, but AC can also show documentation for the currently active candidate.  What happens when you press `F1` in company mode and how is this preferable to displaying it in a popup next to the completion candidates like AC does?

I frequently use this functionality for API discovery, I will search for whatever I'm interested in say `string-` and then `M-n` through the candidates while reviewing the documentation in the popup.

I assume you yourself used auto-complete-mode at some point. What prompted you to assume stewardship of company-mode instead of patching up any annoyances with AC?

**Comment by asmeurer** [comment]
Unless I'm missing it, another huge advantage of auto-complete over company is the [documentation](http://auto-complete.org/doc/manual.html).

**Comment by dgutov** [comment]
I guess so.

I don't know how huge it is, though. For most usage, the homepage and user option descriptions should be sufficient.

**Comment by asmeurer** [comment]
Where are the user option descriptions documented? 

I am trying to switch from auto-complete to company, but I am having a hard time configuring company the way I like, and the lack of docs is not helpful. I was able to find some stuff by searching `M-x customize` for `company`, but it's not enough (e.g., how to I make `TAB` cycle through the completions?).

So I'm not even talking about docs for developers of extensions, it's purely docs for users. The website says almost nothing and README is literally empty.

**Comment by dgutov** [comment]
The website says: `M-x customize-group RET company`. You can read the descriptions there.

> how to I make TAB cycle through the completions?

By writing some Lisp. There no built-in option, but search the issues for instructions.

**Comment by luxchris** [comment]
> I was able to find some stuff by searching M-x customize for company, but it's not enough (e.g., how to I make TAB cycle through the completions?).

Try this.
```
(define-key company-active-map (kbd "TAB") 'company-complete-common-or-cycle)
```
and if you want cycle the other way too, all you must do is copy and paste the function 
`company-complete-common-or-cycle` from *company.el*
```
(defun company-complete-common-or-cycle ()
  "Insert the common part of all candidates, or select the next one."
  (interactive)
  (when (company-manual-begin)
    (let ((tick (buffer-chars-modified-tick)))
      (call-interactively 'company-complete-common)
      (when (eq tick (buffer-chars-modified-tick))
        (let ((company-selection-wrap-around t))
          (call-interactively 'company-select-next))))))
```
and replace the `company-select-next` with `company-select-previous`. Finally bind it to a key
e.g. Shift-TAB 

```
(define-key company-active-map (kbd "<backtab>") 'company-complete-common-or-previous-cycle)
```

**Comment by asmeurer** [comment]
Thanks @luxchris. Would make sense for a negative prefix argument to cycle `company-complete-common-or-cycle` backwards? I'll open a new issue.

**Comment by asmeurer** [comment]
https://github.com/company-mode/company-mode/issues/316

**Comment by dgutov** [comment]
@luxchris Why don't you create a new wiki page with instructions?

**Comment by luxchris** [comment]
@dgutov 
Sorry about that I didn't knew it. I will create a new page at the [wiki](https://github.com/company-mode/company-mode/wiki), ok?

**Comment by dgutov** [comment]
@luxchris Yes please!

**Comment by luxchris** [comment]
@dgutov I just tested your latest  [5e107fd](https://github.com/company-mode/company-mode/commit/5e107fd39629c82ef5c29b549852c2fbda578aaa) commit. Works like a charm! Unfortunately moving forward with *TAB* does not work anymore. Instead it seams like the mini-buffer documentation of the current completion candidate gets updated, however, the key *C-i* works.

But I prefer *C-n* and *C-p* for *next* and *previous* candidate, anyway.

My previous comment [#68](https://github.com/company-mode/company-mode/issues/68#issuecomment-77602328) is now more or less superfluous. To cycle forward or backward someone can now achieve this by defining the following key bindings.

```
(define-key company-active-map (kbd "C-n") (lambda () (interactive) (company-complete-common-or-cycle 1)))
(define-key company-active-map (kbd "C-p") (lambda () (interactive) (company-complete-common-or-cycle -1)))
```

PS.: Should I still write a wiki, at least I/we could create a page with some useful key bindings, like that ones.

Chears, Christian

**Comment by dgutov** [comment]
> Unfortunately moving forward with TAB does not work anymore.

Works for me. Please file an issue with step-by-step repro, and include the beginning of `company-diag` output.

> Should I still write a wiki, at least I/we could create a page with some useful key bindings, like that ones.

If you like. Probably not for this snippet, but anytime you want to post a bunch of code, a wiki page (or a pull request, actually) would likely be better.

**Comment by Alexander-Miller** [comment]
Note: there's now a third-party package to add doc string popups to company: [company-quickhelp](https://github.com/expez/company-quickhelp).

Better yet: it uses pos-tip instead of popup.el, meaning it actually works well.

**Comment by dgutov** [comment]
Thanks for the reminder, updated.

**Comment by jojojames** [comment]
Might not be the best place for this question but I thought it was relevant.

Is there a way in company to replicate how auto-complete initiates completions?
(def example-func() ;z;z)
exam| <-- cursor will hint the rest of the completion and only after a delay show the completion popup. Before the popup shows, user can manually complete the 'hinted completion' and move along.

Maybe a video would be better to describe this, but that's another thing I noticed between the two. (I believe company does something similar to the above but only when there's one completion candidate available).

**Comment by dgutov** [comment]
@jojojames You're right, it's another difference. See https://github.com/company-mode/company-mode/issues/123.

**Comment by jcs090218** [comment]
@dgutov Thanks for the `auto-complete` and `company` comparison, this helps a lot!

**Comment by zw963** [comment]
It is 2020 now, it's been six years since this topic first talk about, any update on this? i get this link from official home page, maybe expect to see some news compare results, thank you!

**Comment by dgutov** [comment]
In that time, we got child frame based frontends such as https://github.com/tumashu/company-posframe/ and https://github.com/sebastiencs/company-box. And some new backends, all third-party ones. Could be forgetting something as well.

Not such there's much point in refreshing this comparison, though: the auto-complete package is not developed anymore.

**Comment by ArneBab** [comment]
> Not such there's much point in refreshing this comparison, though: the auto-complete package is not developed anymore.

This isn’t quite true anymore. In recent years there was development again  and new maintainers are coming in: https://github.com/auto-complete/auto-complete/issues/509

**Comment by dgutov** [comment]
Yes, well. It doesn't seem like there's much "forward" development.

But it's being kept from bitrotting, I guess, which is good for existing useres.
